[WIP] fix clear_host_errors

s-hertel opened the pull request:
##### SUMMARY
Fixes that following `meta: clear_host_errors` failed hosts are not able to get subsequent tasks (due to run_state being ITERATING_COMPLETE) and unreachable hosts getting out of sync (due to being any arbitrary number of tasks behind).

Fixes #35086

Example playbook this fixes:
```
---
- hosts: localhost
  connection: local
  gather_facts: no
  tasks:
  - add_host:
      name: "{{item}}"
      ansible_connection: local
    loop:
    - host0
    - host1

- hosts: host0,host1
  gather_facts: no
  tasks:
    - name: "check host status"
      ping:
        data: success
      failed_when: inventory_hostname == "host0"
      register: ping_result

    - meta: clear_host_errors

    - name: "group by host_status"
      group_by:
        key: "host_status_{{ ping_result.ping | default('fail') }}"
```

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`lib/ansible/plugins/strategy/__init__.py`

s-hertel commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -39,19 +39,55 @@ class HostState:
+            'pending_setup', 'tasks_child_state', 'rescue_child_state', 'always_child_state', 'did_start_at_task',
+        )
+
+        for class_property in class_properties:
```
Maybe it's cleaner to add as
```
@property
def run_state(self):
    ...

@run_state.setter
def run_state(self, value)
    ...
```
etc, but felt redundant

acozine commented:
Docs portion looks good. Thanks @s-hertel.

s-hertel committed:
Save a _restore_state attribute on HostState when marking a host as failed

Forward failed and unreachable hosts during _clear_host_errors to the clear_host_errors task

s-hertel committed:
Clarify documentation

s-hertel committed:
Add tests for clear_host_errors

ci_complete

s-hertel commented:
The docs portion is the same, just moved the changes to the relocated files. The code is better now though - I'm not sure why I over-complicated it. Now PlayIterator just saves a copy of the HostState before marking the host as failed.

s-hertel committed:
Changelog

s-hertel committed:
Oops

ci_complete

bcoca commented:
This shouldn't be the case, but we might need to be mindful that someone somewhere is probably using the current flawed behavior and have a way to reinstate it 'clear_host_errors_for_next_play' or just have them move this meta: call as first pre_task in those?

s-hertel committed:
Save the HostState on the Host object instead since it persists between plays

Add guard rails for strategy plugins not marking hosts as failed via the PlayIterator

ci_complete

s-hertel committed:
Check the PlayIterator as the source of truth for failed hosts instead of TQM's _failed_hosts list.

Include breaking_changes fragment

ci_complete

s-hertel committed:
Add a test for clearing host errors across plays

ci_complete

s-hertel committed:
Reset the restore_state after restoring host to the last good state

ci_complete

mkrizek commented on lib/ansible/plugins/strategy/__init__.py:
```diff
@@ -1148,11 +1148,39 @@ def _evaluate_conditional(h):
+                    # Iterate over tasks until we get to clear_host_errors
+                    s, t = iterator.get_next_task_for_host(host, peek=False)
+                    while t is not None and t.args.get('_raw_params') != 'clear_host_errors':
+                        iterator._host_states[host.name].run_state = s.run_state
```
`iterator.get_next_task_for_host` with `peek=False` already does `iterator._host_states[host.name] = s` so isn't this a noop?

s-hertel replied:
Yes, good catch!

s-hertel commented on lib/ansible/plugins/strategy/__init__.py:
```diff
@@ -1148,11 +1148,39 @@ def _evaluate_conditional(h):
+                    # Iterate over tasks until we get to clear_host_errors
+                    s, t = iterator.get_next_task_for_host(host, peek=False)
+                    while t is not None and t.args.get('_raw_params') != 'clear_host_errors':
+                        iterator._host_states[host.name].run_state = s.run_state
```
```suggestion

```

s-hertel committed:
Update lib/ansible/plugins/strategy/__init__.py

mkrizek commented on lib/ansible/plugins/strategy/__init__.py:
```diff
@@ -1148,11 +1148,38 @@ def _evaluate_conditional(h):
+
+                    # Iterate over tasks until we get to clear_host_errors
+                    s, t = iterator.get_next_task_for_host(host, peek=False)
+                    while t is not None and t.args.get('_raw_params') != 'clear_host_errors':
```
There is a problem when `clear_host_errors` wouldn't be found even if it existed.

Consider following play:
```yaml
- hosts: h1, h2
  gather_facts: no
  tasks:
    - debug:

    - fail:
      when: inventory_hostname == 'h1'

    - block:
      - fail:
      rescue:
        - meta: clear_host_errors

    - debug:
```

`clear_host_errors` is executed in the rescue section on host `h2` so the last `debug` task should execute on both hosts but it is actually executed only on `h2`.

It is because only way `PlayIterator` gets into `rescue` state is by the host being marked as `failed` which is done in the strategy. Here we only iterate through the tasks without the context strategies provide. So when `PlayIterator` gets to the `fail` task in the `block..rescue` it does not know the task fails and so it skips the `rescue` section and does not find the `clear_hosts_errors`, iterating through until it reaches the end of the play.

Obviously if you replace `rescue` by `always` in the above play, it works correctly because `PlayIterator` would iterate through `always` section.

Ugh, hopefully it makes sense.

s-hertel replied:
Ugh, yes. An alternative I had considered was continuing to iterate over failed hosts so clear_host_errors could just bring them back into play where they are. It's a little more invasive, isn't automatically fixed for 3rd party strategy plugins, and I'm not sure that `clear_host_errors` could be compatible with `free` anymore.  ~And it doesn't solve keeping linear tasks in lockstep/duplicate task banners...~ Thanks so much for pondering this one. I'm not sure if there's a great solution here, but I'll keep thinking about it.

s-hertel commented:
I wanted to keep this for future reference so I opened #75394 with the new approach.
